pkg/sql/compile/tools.go

```diff
@@ -0,0 +1,206 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compile
+
+import (
+	"fmt"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/connector"
+	"github.com/matrixorigin/matrixone/pkg/vm"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+	"strings"
+)
+
+var instructionNames = map[int]string{
+	vm.Top:          "top",
+	vm.Join:         "join",
+	vm.Semi:         "semi",
+	vm.Left:         "left",
+	vm.Limit:        "limit",
+	vm.Merge:        "merge",
+	vm.Order:        "order",
+	vm.Group:        "group",
+	vm.Output:       "output",
+	vm.Offset:       "offset",
+	vm.Product:      "product",
+	vm.Restrict:     "restrict",
+	vm.Dispatch:     "dispatch",
+	vm.Connector:    "connect",
+	vm.Projection:   "projection",
+	vm.Anti:         "anti",
+	vm.Single:       "single",
+	vm.Mark:         "mark",
+	vm.LoopJoin:     "loop join",
+	vm.LoopLeft:     "loop left",
+	vm.LoopSemi:     "loop semi",
+	vm.LoopAnti:     "loop anti",
+	vm.LoopSingle:   "loop single",
+	vm.MergeTop:     "merge top",
+	vm.MergeLimit:   "merge limit",
+	vm.MergeOrder:   "merge order",
+	vm.MergeGroup:   "merge group",
+	vm.MergeOffset:  "merge offset",
+	vm.Deletion:     "delete",
+	vm.Insert:       "insert",
+	vm.Update:       "update",
+	vm.External:     "external",
+	vm.Minus:        "minus",
+	vm.Intersect:    "intersect",
+	vm.IntersectAll: "intersect all",
+	vm.HashBuild:    "hash build",
+}
+
+var magicNames = map[int]string{
+	Merge:          "Merge",
+	Normal:         "Normal",
+	Remote:         "Remote",
+	Parallel:       "Parallel",
+	Pushdown:       "Pushdown",
+	CreateDatabase: "CreateDatabase",
+	CreateTable:    "CreateTable",
+	CreateIndex:    "CreateIndex",
+	DropDatabase:   "DropDatabase",
+	DropTable:      "DropTable",
+	DropIndex:      "DropIndex",
+	Deletion:       "Deletion",
+	Insert:         "Insert",
+	Update:         "Update",
+	InsertValues:   "InsertValues",
+}
+
+var _ = ShowScopes
+
+// ShowScopes show information of a scope structure.
+func ShowScopes(ss []*Scope) string {
+	receiverMap := make(map[*process.WaitRegister]int)
+	for i := range ss {
+		generateReceiverMap(ss[i], receiverMap)
+	}
+
+	return showScopes(ss, 0, receiverMap)
+}
+
+func magicShow(magic int) string {
+	name, ok := magicNames[magic]
+	if ok {
+		return name
+	}
+	return "unknown"
+}
+
+func showInstruction(instruction vm.Instruction, mp map[*process.WaitRegister]int) string {
+	id := instruction.Op
+	name, ok := instructionNames[id]
+	if ok {
+		str := name
+		if id == vm.Connector {
+			var receiver = "unknown"
+			arg := instruction.Arg.(*connector.Argument)
+			if receiverId, okk := mp[arg.Reg]; okk {
+				receiver = fmt.Sprintf("%d", receiverId)
+			}
+			str += fmt.Sprintf(" to MergeReceiver %s", receiver)
+		}
+		return str
+	}
+	return "unknown"
+}
+
+func getDataSource(source *Source) string {
+	if source == nil {
+		return "nil"
```

Comment:
What is this?   Later you will parse "nil"?

Reply:
You should not be dealing with pretty print,

Reply:
you're right. 
I have set these functions as anonymous functions, it will only be called at debug functions.